Add database

savannavea opened the pull request.

savannavea committed:
"Финальный проект спринта 11 "

savannavea committed:
"Финальный проект спринта 11 "

savannavea committed:
"Финальный проект спринта 11 "

savannavea committed:
"Финальный проект спринта 11 "

FanaticFiz commented on src/main/java/ru/yandex/practicum/controller/GenreController.java:
```diff
@@ -0,0 +1,32 @@
+
+    @GetMapping("/{id}")
+    public Genre findById(@PathVariable int id) {
+        return genreService.getGenreById(id);
```
Переименуй в `getById`.
Упоминание жанра в названии метода излишне.

FanaticFiz commented on src/main/java/ru/yandex/practicum/controller/MpaController.java:
```diff
@@ -0,0 +1,33 @@
+
+    @GetMapping("/{id}")
+    public MPA getById(@PathVariable int id) {
+        return mpaService.getMPAById(id);
```
`getMPAById` - это то же переименовать, но тут допущена и другая ошибка.
`MPA -> Mpa `- кемелКейс

FanaticFiz commented on src/main/java/ru/yandex/practicum/controller/UserController.java:
```diff
@@ -82,13 +82,13 @@ public void deleteFriends(@PathVariable int id, @PathVariable int friendId) {
     @GetMapping("/{id}/friends")
     @ResponseStatus(HttpStatus.OK)
-    public List<Optional<User>> getFriendsList(@PathVariable int id) {
+    public Optional<List<User>> getFriendsList(@PathVariable int id) {
```
List не нуждается в обертке Optional - нужно всегда возвращать пустой список, если ничего не нашлось.

FanaticFiz commented on src/main/java/ru/yandex/practicum/model/Film.java:
```diff
@@ -31,13 +30,23 @@ public class Film {
+    @JsonProperty
+    private final List<Genre> genres = new ArrayList<>();
+
+    public Map<String, Object> toMap() {
```
Используй отдельный класс, ентити должна оставаться простой

FanaticFiz commented on src/main/java/ru/yandex/practicum/service/UserService.java:
```diff
@@ -15,12 +15,11 @@ public interface UserService {
-    List<Optional<User>> getFriendsList(int id);
-
-    List<Optional<User>> getCommonFriends(int userId, int friendId);
+    Optional<List<User>> getFriendsList(int id);
```
`getFriendsList -> getFriends`

FanaticFiz commented on src/main/java/ru/yandex/practicum/service/impl/GenreServiceImpl.java:
```diff
@@ -0,0 +1,32 @@
+    @Override
+    public List<Genre> getAll() {
+        return genreStorage.findAll().stream()
+                .sorted(Comparator.comparingInt(Genre::getId))
```
Зачем сортировать "вручную", сделай это запросом в БД

FanaticFiz commented on src/main/java/ru/yandex/practicum/storage/impl/InMemoryFilmStorage.java:
```diff
@@ -1,5 +1,5 @@
 package ru.yandex.practicum.storage.impl;
-
+/*
```
Удаляй эти реализации, зачем же их комментить...

FanaticFiz replied:
Не понимаю почему ты вернула работу не проработав несколько комментариев в коде? 
В пачку с вопросами тоже не приходила...

FanaticFiz commented on src/main/java/ru/yandex/practicum/storage/impl/FriendshipDbStorage.java:
```diff
@@ -0,0 +1,45 @@
+
+    @Override
+    public boolean addFriend(int userId, int friendId) {
+        String sql = "insert into friendship(user_id, friend_id, status) " +
```
Оформляй запросы, набирая команды SQL в верхнем регистре: SELECT

FanaticFiz commented on src/main/java/ru/yandex/practicum/storage/impl/GenreDbStorage.java:
```diff
@@ -0,0 +1,57 @@
+    }
+
+    private Genre makeGenre(ResultSet rs) throws SQLException {
+        return Genre.builder().id(rs.getInt("genre_id"))
```
Каждая новая операция на новой строке

```
    private Genre makeGenre(ResultSet rs) throws SQLException {
        return Genre.builder()
                    .id(rs.getInt("genre_id"))
                    .name(rs.getString("name"))
                    .build();
    }
```

FanaticFiz commented on src/main/java/ru/yandex/practicum/storage/impl/FilmDbStorage.java:
```diff
@@ -0,0 +1,135 @@
+    @Override
+    public List<Film> findAll() {
+        String sql = "select * from films left join mpa on films.mpa_id = mpa.mpa_id;";
+        return jdbcTemplate.query(sql, (rs, rowNum) -> mapRowToFilm(rs));
```
Тут нужно быть аккуратнее!

Вот так в лоб нельзя делать в данном случае!
Давай поясню...
пусть у нас есть 100 фильмов... а у каждого из них есть жанры
для того чтобы собрать всё в одно и отдать при запросе нам нужно:

- достать все фильмы
- и все жанры к ним

Если делать "в лоб" то приходиться к каждому из 100-а фильмов отдельным запросом доставать еще и жанры...
получается ОДИН запрос на получение всех 100 фильмов и еще 100 запросов чтобы к каждому фильму подтянуть жанры....

Это называется проблема `N+1`
Чтобы её избежать мы должны заморочиться и в итоге приходим к всего ДВУМ запросам, вместо 101 !

Первый запрос на все фильмы, второй на все жанры для этих фильмов.

И все эти манипуляции следует выполнять в сервисе!

FanaticFiz commented on src/main/resources/application.properties:
```diff
@@ -0,0 +1,6 @@
+spring.sql.init.mode=always
```
Этот файл в кодировке `ISO-8859-1`, а должно быть `UTF-8`

FanaticFiz commented on src/main/resources/schema.sql:
```diff
@@ -0,0 +1,66 @@
+
+create table IF NOT EXISTS FILMS
+(
+    FILM_ID      INTEGER auto_increment
```
- Желательно бы в нижнем регистре, так принято в БД
- Не нужно дублировать у таблички `film` есть поле `id`. А называть `film_id` избыточно!
- Причесать файл, выровнять

```
create table IF NOT EXISTS FILMS
(
    FILM_ID      INTEGER auto_increment
        primary key,
    NAME         CHARACTER VARYING(150) not null,
    DESCRIPTION  CHARACTER VARYING(200) not null,
    RELEASE_DATE DATE                   not null,
    DURATION     INTEGER                not null,
    MPA_ID       INTEGER                not null
        references MPA (MPA_ID) ON DELETE RESTRICT
);
```

FanaticFiz reviewed (changes requested):
Финальный комментарий на Я.Практикуме

savannavea committed:
"Финальный проект спринта 11. Исправления 1.1 "

savannavea committed:
"Финальный проект спринта 11. Исправления 1.2 "

savannavea committed:
"Финальный проект спринта 11. Исправления 1
(В FilmStorage.findAll(Films), в sql одним запросом вытащила мпа, жанры и лайки)"

FanaticFiz commented on src/main/resources/schema.sql:
```diff
@@ -0,0 +1,55 @@
+
+CREATE TABLE IF NOT EXISTS films
+(
+    id INTEGER auto_increment primary key,
```
С выравниванием куда удобнее читать

```
CREATE TABLE IF NOT EXISTS films
(
    id           INTEGER auto_increment primary key,
    name         CHARACTER VARYING(150) NOT NULL,
    description  CHARACTER VARYING(200) NOT NULL,
    release_date DATE                   NOT NULL,
    duration     INTEGER                NOT NULL,
    mpa_id       INTEGER                NOT NULL REFERENCES mpa (mpa_id) ON DELETE RESTRICT
);
```

FanaticFiz commented on src/main/java/ru/yandex/practicum/storage/impl/FilmDbStorage.java:
```diff
@@ -0,0 +1,151 @@
+                "LEFT JOIN mpa ON films.mpa_id = mpa.mpa_id " +
+                "LEFT JOIN film_genre ON films.id = film_genre.film_id " +
+                "LEFT JOIN likes ON films.id = likes.film_id;";
+        return jdbcTemplate.query(sql, (rs, rowNum) -> mapRowToFilm(rs));
```
У тебя в `mapRowToFilm` есть вызов `genreDbStorage.findGenresByFilmId(id)` в котором происходит выборка жанров для каждого фильма отдельно! 

Добавь логирования может, чтобы подсветить проблему.

FanaticFiz commented on src/main/java/ru/yandex/practicum/storage/MpaStorage.java:
```diff
@@ -0,0 +1,13 @@
+
+public interface MpaStorage {
+
+    Optional<Mpa> findMpaById(int id);
```
findMpaById -> findById

Квест - найди все похожие места! ))

FanaticFiz commented on src/main/java/ru/yandex/practicum/storage/GenreStorage.java:
```diff
@@ -0,0 +1,15 @@
+
+public interface GenreStorage {
+
+    Optional<Genre> findGenreById(int id);
```
findGenreById -> findById

FanaticFiz commented on src/main/java/ru/yandex/practicum/service/impl/UserServiceImpl.java:
```diff
@@ -45,61 +46,52 @@ public User getUserById(int id) {
+    public List<User> getFriends(int userId) {
+        List<User> friendsList = new ArrayList<>();
+
+        for (Integer friend : friendshipStorage.findFriendsByUser(userId)) {
```
Тут ты тоже злоупотребляешь циклом! 
Выбери одним запросом.

И еще. Обрати внимание на `findUserById(friend).get()` ты не можешь брать `get` не проверив что Optional содержит значение!

savannavea committed:
"Финальный проект спринта 11. Исправления 2. Удалила старые классы и список жанров отсортировала запросом в БД"

savannavea committed:
"Финальный проект спринта 11. Исправления 2. Подровняла проект."

savannavea committed:
"Финальный проект спринта 11. Исправления 2."
